### ISSUE TITLE
When calling torch.cumprod on a float16 tensor, the CPU and CUDA implementations produce different outputs.

### ISSUE THREAD

**Issue by atester353** [pytorch/pytorch#156018]
### 🐛 Describe the bug

```python
import torch
import numpy as np
np_array = np.array([[[6.5, 6.5, 6.5, 6.5, 6.5],
                [6.5, 6.5, 6.5, 6.5, 6.5],
                [6.5, 6.5, 6.5, 6.5, 6.5],
                [6.5, 6.5, 6.5, 6.5, 6.5]],
        
               [[6.5, 6.5, 6.5, 6.5, 6.5],
                [6.5, 6.5, 6.5, 6.5, 6.5],
                [6.5, 6.5, 6.5, 6.5, 6.5],
                [6.5, 6.5, 6.5, 6.5, 6.5]],
      
               [[6.5, 6.5, 6.5, 6.5, 6.5],
                [6.5, 6.5, 6.5, 6.5, 6.5],
                [6.5, 6.5, 6.5, 6.5, 6.5],
                [6.5, 6.5, 6.5, 6.5, 6.5]]], dtype=np.float16)
input = torch.from_numpy(np_array)
dim = 1
out_gpu = torch.cumprod(input.cuda(),dim)
out_cpu = torch.cumprod(input, dim)
print(out_cpu)
print(out_gpu.cpu())    

# CPU result
# → tensor([[[   6.5000,    6.5000,    6.5000,    6.5000,    6.5000],
         [  42.2500,   42.2500,   42.2500,   42.2500,   42.2500],
         [ 274.5000,  274.5000,  274.5000,  274.5000,  274.5000],
         [1785.0000, 1785.0000, 1785.0000, 1785.0000, 1785.0000]],

        [[   6.5000,    6.5000,    6.5000,    6.5000,    6.5000],
         [  42.2500,   42.2500,   42.2500,   42.2500,   42.2500],
         [ 274.5000,  274.5000,  274.5000,  274.5000,  274.5000],
         [1785.0000, 1785.0000, 1785.0000, 1785.0000, 1785.0000]],

        [[   6.5000,    6.5000,    6.5000,    6.5000,    6.5000],
         [  42.2500,   42.2500,   42.2500,   42.2500,   42.2500],
         [ 274.5000,  274.5000,  274.5000,  274.5000,  274.5000],
         [1785.0000, 1785.0000, 1785.0000, 1785.0000, 1785.0000]]],
       dtype=torch.float16)
# GPU result
# → tensor([[[   6.5000,    6.5000,    6.5000,    6.5000,    6.5000],
         [  42.2500,   42.2500,   42.2500,   42.2500,   42.2500],
         [ 274.5000,  274.5000,  274.5000,  274.5000,  274.5000],
         [1784.0000, 1784.0000, 1784.0000, 1784.0000, 1784.0000]],

        [[   6.5000,    6.5000,    6.5000,    6.5000,    6.5000],
         [  42.2500,   42.2500,   42.2500,   42.2500,   42.2500],
         [ 274.5000,  274.5000,  274.5000,  274.5000,  274.5000],
         [1784.0000, 1784.0000, 1784.0000, 1784.0000, 1784.0000]],

        [[   6.5000,    6.5000,    6.5000,    6.5000,    6.5000],
         [  42.2500,   42.2500,   42.2500,   42.2500,   42.2500],
         [ 274.5000,  274.5000,  274.5000,  274.5000,  274.5000],
         [1784.0000, 1784.0000, 1784.0000, 1784.0000, 1784.0000]]],
       dtype=torch.float16)

[+]### Versions[/+]
[+]PyTorch version: 2.7.0+cu126[/+]


### Versions

I'm not sure whether this numerical gap is allowed

**Comment by huiyan2021** [comment]
Please refer to: https://docs.pytorch.org/docs/stable/notes/numerical_accuracy.html

**Comment by bdhirsh** [comment]
This looks like expected numerical differences given that you're using float16

**Comment by nobody81127** [comment]
I applied a floating-point error analysis algorithm to calculate the error bounds for the result in cuda, and I found that the result in cpu falls within this range. Therefore, I believe this is indeed a floating-point error, likely caused by differences in the order of operations. @atester353 @bdhirsh
